cloud/WFLY-11824_wildfly_operator_kubernetes.adoc

```diff
@@ -0,0 +1,126 @@
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-11824[WFLY-11824]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1149[EAP7-1149]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+ * TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+[ ] QE
+
+=== Affected Projects or Components
+
+ * https://github.com/wildfly/wildfly[WildFly]
+ * https://github.com/openshift-s2i/s2i-wildfly[WildFly s2i]
+
+
+=== Other Interested Projects
+
+ * https://github.com/jboss-dockerfiles/wildfly[WildFly Docker image]
+ * Keycloak / Red Hat SSO
+ * RHPAM
+
+== Requirements
+
+
+=== Hard Requirements
+
+==== Create a Custom Resource Definition (CRD) for WildFly.
+
+Provide a Custom Resource Definition (CRD) for an application based on a WildFly application server.
+
+The CRD will allow the user to:
+
+* Specify the name an application image that incorporates a WildFly server and an deployment that will be deployed by that server.
+** The name can include either a fixed or floating tag.
+* Specify an optional ConfigMap where the WildFly standalone configuration can be read from.
+* Specify the desired number of replicas of that image
+* Specify whether the replicas are intended to form a High Availability cluster (default is true).
+* Specify whether the containers are intended to be 'stateful'
+** default is to be determined (might be inferred from the server configuration)
+* Specify information for interfaces for which a Service and Route should be established
+** perhaps expose 8080 by default with an option to disable that
+** same for 9990 to expose management interface (including health check and metrics endpoints)
+* Specify information for interfaces for which a headless service should be established.
+
+### Operator requirements
+
+The WildFly operation will provide the following features:
+
+* In reaction to deployment of a Custom Resource with the CRD's type start up a cluster of containers running the specified image, with the desired number of replicas.
+* If the containers are meant to form a High Availability cluster, perform scale up and scale down in a manner designed to optimize cluster efficiency
+** Start one container and then start the others, thus ensuring the later nodes see the first as the group coordinator, avoiding mergers.
+* If the resources are configured as stateful
+** Ensure the existence of the necessary persistent volume(s) and persistent volume claims necessary for each container to have access to its own writable persistence
+** If the necessary facilities to allow this are not available, container start should be aborted
+** Appropriate transaction recovery activities should be performed
+* Establish the appropriate services and routes
+* Detect new versions of the image in the container catalog and take appropriate action. Possible actions include
```

Comment:
This should not happen automatically, but should only happen by the user triggering a change in the version of the CR defining the application. Automatic deployments based on an image available on a registry are only acceptable in certain development scenarios.